book/src/implemented-proposals/repair-service.md

```diff
@@ -2,48 +2,100 @@
+     Note: that once these empty slots are added to blockstore, the 
+     `Shred Repair` protocol should attempt to fill those slots.
 
-     Note: Validators will only accept responses containing shreds within the current verifiable epoch \(epoch the validator has a leader schedule for\).
-3. Repairmen \(Addresses Challenge \#3\): This part of the repair protocol is the primary mechanism by which new nodes joining the cluster catch up after loading a snapshot. This protocol works in a "forward" fashion, so validators can verify every shred that they receive against a known leader schedule.
+     Note: Validators will only accept responses containing shreds within the
+     current verifiable epoch \(epoch the validator has a leader schedule 
+     for\).
 
-   Each validator advertises in gossip:
+Validators should try to send orphan requests to validators who have marked that
+orphan as completed in their EpochSlots. If no such validators exist, then
+randomly select a validator in a stake-weighted fashion.
 
-   * Current root
-   * The set of all completed slots in the confirmed epochs \(an epoch that was calculated based on a bank &lt;= current root\) past the current root
+## Repair Response Protocol
 
-     Observers of this gossip message with higher epochs \(repairmen\) send shreds to catch the lagging node up with the rest of the cluster. The repairmen are responsible for sending the slots within the epochs that are confrimed by the advertised `root` in gossip. The repairmen divide the responsibility of sending each of the missing slots in these epochs based on a random seed \(simple shred.index iteration by N, seeded with the repairman's node\_pubkey\). Ideally, each repairman in an N node cluster \(N nodes whose epochs are higher than that of the repairee\) sends 1/N of the missing shreds. Both data and coding shreds for missing slots are sent. Repairmen do not send shreds again to the same validator until they see the message in gossip updated, at which point they perform another iteration of this protocol.
+When a validator receives a request for a shred `S`, they respond with the
+shred if they have it. 
 
-     Gossip messages are updated every time a validator receives a complete slot within the epoch. Completed slots are detected by blockstore and sent over a channel to RepairService. It is important to note that we know that by the time a slot X is complete, the epoch schedule must exist for the epoch that contains slot X because WindowService will reject shreds for unconfirmed epochs. When a newly completed slot is detected, we also update the current root if it has changed since the last update. The root is made available to RepairService through Blockstore, which holds the latest root.
+When a validator receives a shred through a repair response, they check 
+`EpochSlots` to see if <= `1/3` of the network has marked this slot as 
```

Comment:
When a validator receives a request, it could prioritize the response with retransmit shreds if less than 1/3rd of the network has the block.

Reply:
Receiving validators must retransmit once and only once.

Reply:
@aeyakovenko receiving validators as in validators that receive the request or the response?

Reply:
Alice sends the repair request to Bob with a range of missing shreds.

If Bob sees that 1/3- have the slot, he can prioritize shreds that Alice would retransmit.